numba/cuda/tests/cudapy/test_ufuncs.py

```diff
@@ -0,0 +1,144 @@
+             types.Array(types.complex64, 1, 'C')),
+            (np.array([-0.5 - 0.5j, 0.0, 0.5 + 0.5j], dtype='c16'),
+             types.Array(types.complex128, 1, 'C')),
+        ])
+
+        self._low_occupancy_warnings = config.CUDA_LOW_OCCUPANCY_WARNINGS
+        self._warn_on_implicit_copy = config.CUDA_WARN_ON_IMPLICIT_COPY
+
+        # Disable warnings about low gpu utilization in the test suite
+        config.CUDA_LOW_OCCUPANCY_WARNINGS = 0
+        # Disable warnings about host arrays in the test suite
+        config.CUDA_WARN_ON_IMPLICIT_COPY = 0
+
+    def tearDown(self):
+        # Restore original warning settings
+        config.CUDA_LOW_OCCUPANCY_WARNINGS = self._low_occupancy_warnings
+        config.CUDA_WARN_ON_IMPLICIT_COPY = self._warn_on_implicit_copy
+
+    def _make_ufunc_usecase(self, ufunc):
+        return _make_ufunc_usecase(ufunc)
+
+    @functools.lru_cache(maxsize=None)
+    def _compile(self, pyfunc, args):
+        # We return an already-configured kernel so that basic_ufunc_test can
+        # call it just like it does for a CPU function
+        return cuda.jit(args)(pyfunc)[1, 1]
+
+    ############################################################################
+    # Trigonometric Functions
+
+    def test_sin_ufunc(self):
+        self.basic_ufunc_test(np.sin, kinds='cf')
+
+    def test_cos_ufunc(self):
+        self.basic_ufunc_test(np.cos, kinds='cf')
+
+    def test_tan_ufunc(self):
+        self.basic_ufunc_test(np.tan, kinds='cf')
+
+    def test_arcsin_ufunc(self):
+        self.basic_ufunc_test(np.arcsin, kinds='cf')
+
+    def test_arccos_ufunc(self):
+        self.basic_ufunc_test(np.arccos, kinds='cf')
+
+    def test_arctan_ufunc(self):
+        self.basic_ufunc_test(np.arctan, kinds='cf')
+
+    def test_arctan2_ufunc(self):
+        self.basic_ufunc_test(np.arctan2, kinds='f')
+
+    def test_hypot_ufunc(self):
+        self.basic_ufunc_test(np.hypot, kinds='f')
+
+    def test_sinh_ufunc(self):
+        self.basic_ufunc_test(np.sinh, kinds='cf')
+
+    def test_cosh_ufunc(self):
+        self.basic_ufunc_test(np.cosh, kinds='cf')
+
+    def test_tanh_ufunc(self):
+        self.basic_ufunc_test(np.tanh, kinds='cf')
+
+    def test_arcsinh_ufunc(self):
+        self.basic_ufunc_test(np.arcsinh, kinds='cf')
+
+    def test_arccosh_ufunc(self):
+        self.basic_ufunc_test(np.arccosh, kinds='cf')
+
+    def test_arctanh_ufunc(self):
+        # arctanh is only valid is only finite in the range ]-1, 1[
+        # This means that for any of the integer types it will produce
+        # conversion from infinity/-infinity to integer. That's undefined
+        # behavior in C, so the results may vary from implementation to
+        # implementation. This means that the result from the compiler
+        # used to compile NumPy may differ from the result generated by
+        # llvm. Skipping the integer types in this test avoids failed
+        # tests because of this.
```

Comment:
That seems to be the case:

```
In [8]: np.arctanh(np.int32(1))
<ipython-input-8-ad15ffdef1f4>:1: RuntimeWarning: divide by zero encountered in arctanh
  np.arctanh(np.int32(1))
Out[8]: inf

In [9]: np.arctanh(np.int32(-1))
<ipython-input-9-a4fc10f112e9>:1: RuntimeWarning: divide by zero encountered in arctanh
  np.arctanh(np.int32(-1))
Out[9]: -inf
```

Reply:
Do you have a view on what to do with these test input restrictions and commentary given they are potentially obsolete? I think given it's late in the release cycle, perhaps leave this for now as the impact is near zero and revisit when working on #8700?

Reply:
I don't think these are obsolete - I think I concretely get different answers between the CPU and CUDA target, because the underlying libraries have different implementation choices that they were free to make. Either way, I would indeed like to leave this for now and revisit as part of #8700.